From 58ca160b769537b04586152cddb1acd600d11b19 Mon Sep 17 00:00:00 2001
From: Leo Martin <contact@leomartin.ch>
Date: Tue, 25 Nov 2025 23:27:07 +0100
Subject: [PATCH] Patch to add a border for wmenu in center mode

---
 menu.c    | 18 +++++++++++++-----
 menu.h    |  1 +
 render.c  | 27 ++++++++++++++++++---------
 wayland.c |  2 +-
 4 files changed, 33 insertions(+), 15 deletions(-)

diff --git a/menu.c b/menu.c
index 2a333fa..bb3f49e 100644
--- a/menu.c
+++ b/menu.c
@@ -9,7 +9,7 @@
 #include <string.h>
 #include <strings.h>
 #include <time.h>
-#include <unistd.h>
+#include <getopt.h>
 #include <sys/mman.h>
 #include <sys/timerfd.h>
 #include <wayland-client.h>
@@ -88,9 +88,17 @@ void menu_getopts(struct menu *menu, int argc, char *argv[]) {
 		"Usage: wmenu [-biPv] [-f font] [-l lines] [-o output] [-p prompt]\n"
 		"\t[-N color] [-n color] [-M color] [-m color] [-S color] [-s color]\n";
 
+	static struct option long_options[] = {
+		{"bw", required_argument, NULL, 1000},
+		{NULL, 0, NULL, 0}
+	};
+
 	int opt;
-	while ((opt = getopt(argc, argv, "bhiPcvf:l:o:p:N:n:M:m:S:s:")) != -1) {
+	while ((opt = getopt_long(argc, argv, "bhiPcvf:l:o:p:N:n:M:m:S:s:", long_options, NULL)) != -1) {
 		switch (opt) {
+		case 1000: // --bw
+			menu->border_width = atoi(optarg);
+			break;
 		case 'b':
 			menu->bottom = true;
 			break;
@@ -164,12 +172,12 @@ void menu_getopts(struct menu *menu, int argc, char *argv[]) {
 void menu_calc_height(struct menu *menu) {
 	int height = get_font_height(menu->font);
 	menu->line_height = height + 2;
-	menu->height = menu->line_height;
+	menu->height = menu->line_height + (2 * menu->border_width);
 	if (menu->lines > 0) {
 		if (menu->item_count < (size_t)menu->lines) {
 			menu->lines = menu->item_count;
 		}
-		menu->height += menu->height * menu->lines;
+		menu->height += menu->line_height * menu->lines;
 	}
 	menu->padding = height / 2;
 }
@@ -257,7 +265,7 @@ static void page_items(struct menu *menu) {
 	} else {
 		// Calculate available space
 		int max_width = menu->width - menu->inputw - menu->promptw
-			- menu->left_arrow - menu->right_arrow;
+			- menu->left_arrow - menu->right_arrow - (2 * menu->border_width);
 
 		struct page *pages_end = NULL;
 		struct item *item = menu->matches;
diff --git a/menu.h b/menu.h
index 1801cac..23847c5 100644
--- a/menu.h
+++ b/menu.h
@@ -66,6 +66,7 @@ struct menu {
 	int width;
 	int height;
 	int line_height;
+	int border_width;
 	int padding;
 	int inputw;
 	int promptw;
diff --git a/render.c b/render.c
index e86d93b..3fd52ad 100644
--- a/render.c
+++ b/render.c
@@ -79,7 +79,7 @@ static void render_prompt(struct menu *menu, cairo_t *cairo) {
 	if (!menu->prompt) {
 		return;
 	}
-	render_text(menu, cairo, menu->prompt, 0, 0, 0,
+	render_text(menu, cairo, menu->prompt, menu->border_width, menu->border_width, 0,
 		menu->promptbg, menu->promptfg, menu->padding, menu->padding/2);
 }
 
@@ -96,7 +96,8 @@ static void render_input(struct menu *menu, cairo_t *cairo) {
 	}
 
 	render_text(menu, cairo, menu->passwd ? censort : menu->input,
-		menu->promptw, 0, 0, 0, menu->normalfg, menu->padding, menu->padding);
+		menu->promptw + menu->border_width, menu->border_width, menu->border_width, 0,
+		menu->normalfg, menu->padding, menu->padding);
 
 	if (censort) {
 		free(censort);
@@ -107,11 +108,11 @@ static void render_input(struct menu *menu, cairo_t *cairo) {
 static void render_cursor(struct menu *menu, cairo_t *cairo) {
 	const int cursor_width = 2;
 	const int cursor_margin = 2;
-	int cursor_pos = menu->promptw + menu->padding
+	int cursor_pos = menu->promptw + menu->padding + menu->border_width
 		+ text_width(cairo, menu->font, menu->input)
 		- text_width(cairo, menu->font, &menu->input[menu->cursor])
 		- cursor_width / 2;
-	cairo_rectangle(cairo, cursor_pos, cursor_margin, cursor_width,
+	cairo_rectangle(cairo, cursor_pos, cursor_margin + menu->border_width, cursor_width,
 			menu->line_height - 2 * cursor_margin);
 	cairo_fill(cairo);
 }
@@ -121,7 +122,7 @@ static int render_horizontal_item(struct menu *menu, cairo_t *cairo, struct item
 	uint32_t bg_color = menu->sel == item ? menu->selectionbg : menu->normalbg;
 	uint32_t fg_color = menu->sel == item ? menu->selectionfg : menu->normalfg;
 
-	return render_text(menu, cairo, item->text, x, 0, 0,
+	return render_text(menu, cairo, item->text, x + menu->border_width, menu->border_width, 0,
 		bg_color, fg_color, menu->padding, menu->padding);
 }
 
@@ -130,8 +131,8 @@ static int render_vertical_item(struct menu *menu, cairo_t *cairo, struct item *
 	uint32_t bg_color = menu->sel == item ? menu->selectionbg : menu->normalbg;
 	uint32_t fg_color = menu->sel == item ? menu->selectionfg : menu->normalfg;
 
-	render_text(menu, cairo, item->text, x, y, menu->width - x,
-		bg_color, fg_color, menu->padding, 0);
+	render_text(menu, cairo, item->text, x + menu->border_width, y + menu->border_width,
+		menu->width - x - (2 * menu->border_width), bg_color, fg_color, menu->padding, 0);
 	return menu->line_height;
 }
 
@@ -144,11 +145,11 @@ static void render_horizontal_page(struct menu *menu, cairo_t *cairo, struct pag
 
 	// Draw left and right scroll indicators if necessary
 	if (page->prev) {
-		cairo_move_to(cairo, menu->promptw + menu->inputw + menu->padding, 0);
+		cairo_move_to(cairo, menu->promptw + menu->inputw + menu->padding + menu->border_width, menu->border_width);
 		pango_printf(cairo, menu->font, 1, "<");
 	}
 	if (page->next) {
-		cairo_move_to(cairo, menu->width - menu->right_arrow + menu->padding, 0);
+		cairo_move_to(cairo, menu->width - menu->right_arrow + menu->padding - menu->border_width, menu->border_width);
 		pango_printf(cairo, menu->font, 1, ">");
 	}
 }
@@ -169,6 +170,14 @@ static void render_to_cairo(struct menu *menu, cairo_t *cairo) {
 	cairo_set_source_u32(cairo, menu->normalbg);
 	cairo_paint(cairo);
 
+	// Render border
+	if (menu->border_width) {
+		cairo_set_line_width (cairo, menu->border_width);
+		cairo_set_source_u32(cairo, menu->selectionbg);
+		cairo_rectangle (cairo, menu->border_width / 2, menu->border_width / 2, menu->width - menu->border_width, menu->height - menu->border_width);
+		cairo_stroke (cairo);
+	}
+
 	// Render prompt and input
 	render_prompt(menu, cairo);
 	render_input(menu, cairo);
diff --git a/wayland.c b/wayland.c
index 46717ac..1224ab3 100644
--- a/wayland.c
+++ b/wayland.c
@@ -460,7 +460,7 @@ int menu_run(struct menu *menu) {
 	if (menu->centered) {
 		anchor = 0;
 		int min_width = 500;
-		wd = MIN(MAX(menu->inputw + menu->promptw, min_width) + 15, screen_info.width);
+		wd = MIN(MAX(menu->inputw + menu->promptw, min_width) + 15 + (2 * menu->border_width), screen_info.width);
 	} else {
 		anchor = (menu->bottom ? ZWLR_LAYER_SURFACE_V1_ANCHOR_BOTTOM : ZWLR_LAYER_SURFACE_V1_ANCHOR_TOP) |
 			ZWLR_LAYER_SURFACE_V1_ANCHOR_LEFT |
-- 
2.52.0

